### ISSUE TITLE
[FR] Use new num_chains arg in cmdstan for dense_e and diag_e

### ISSUE THREAD

**Issue by SteveBronder** [stan-dev/cmdstanr#534]
**Is your feature request related to a problem? Please describe.**

So once https://github.com/stan-dev/cmdstan/pull/987 goes into cmdstan, as long as the user compiled with threading on then we should be able to run multiple chains within one Stan program by specifying `num_chains` inside of the `sample` argument! 

**Describe the solution you'd like**

Is there a reasonable line for cmdstanr to check if the method is hmc with dense or diag_e inverse metric and then when the user specifies parallel chains, internally cmdstanr passes those as command args to cmdstan instead of using R level parallelism? This should save folks a lot of memory and likewise give a nice little speed bump.

This would also have to wait / check that cmdstanr is at least greater than version 2.27 and has compiled with threading on

**Describe alternatives you've considered**

I'm not familiar enough with the internals of cmdstanr to know other alts, though I'm happy to discuss whatever can work for yinz!

**Additional context**
Add any other context about the feature request here.

**Comment by rok-cesnovar** [comment]
@SteveBronder Thanks! I set the milestone for after the 2.27++ release.

Yeah, I think this will just be a special case if version > 2.27, STAN_THREADS is on and the appropriate methods are used then just run a single executable instead of running multiple and waiting for feedback (https://github.com/stan-dev/cmdstanr/blob/master/R/run.R#L344).

So this will work with sampling for both hmc and nuts and for both dense and diag inverse metric?

**Comment by SteveBronder** [comment]
> So this will work with sampling for both hmc and nuts and for both dense and diag inverse metric?

It's just nuts with the dense and diag metric. Once I have all the cmdstan stuff in for nuts then I'm going to go back to the Stan repo and add it for hmc and all the other stuff in sampling

**Comment by rok-cesnovar** [comment]
Yeah, no rush. Was just trying to double check. These 2 are definitely the most used so no rush for the rest.

**Comment by SteveBronder** [comment]
^I'll do this next week

**Comment by rok-cesnovar** [comment]
Blocked by https://github.com/stan-dev/cmdstan/issues/1058 (should be fixed fast and accompanied by a 2.28.2 release).

**Comment by SteveBronder** [comment]
Is this still blocked? @rok-cesnovar if you don't have an impl yet would it be a lot easier if all the NUTS cmdstan calls could use the same num_threads scheme? I think for cmdstanr we only need NUTS with and without adaptation

**Comment by rok-cesnovar** [comment]
Not blocked. I got distracted and forgot… 

I have a semi done branch. Will finish it up this weekend.

**Comment by SteveBronder** [comment]
Update on this, after https://github.com/stan-dev/cmdstan/pull/1176 is merged into cmdstan all of hmc, fixed params, and generated quantites can use the num_chains argument. This should simplify some of the higher level code and is also more performant because instead of making N copies of data / model for each chain we just have one data/model that is shared across all chains

**Comment by jgabry** [comment]
Cool, thanks @SteveBronder

**Comment by jgabry** [comment]
@SteveBronder pinging you as promised to remind you to see if Claude can help us make some progress on this before v1.0!

**Comment by jgabry** [comment]
@SteveBronder pinging you again as requested

**Comment by SteveBronder** [comment]
@jgabry check out the diff below for the version of this I have working so far 

https://github.com/stan-dev/cmdstanr/compare/master...feature/cmstan-internal-parallel
